csharp/src/Apache.Arrow/C/CArrowSchema.cs

```diff
@@ -0,0 +1,584 @@
+            }
+            public void Visit(Decimal256Type datatype)
+            {
+                FormatString = $"w:{datatype.Precision},{datatype.Scale},256";
+            }
+            // Date
+            public void Visit(Date32Type _) => FormatString = "tdD";
+            public void Visit(Date64Type _) => FormatString = "tdm";
+
+            private char TimeUnitComponent(TimeUnit unit) => unit switch
+            {
+                TimeUnit.Second => 's',
+                TimeUnit.Millisecond => 'm',
+                TimeUnit.Microsecond => 'u',
+                TimeUnit.Nanosecond => 'n',
+                _ => throw new InvalidDataException($"Unsupported time unit for export: {unit}"),
+            };
+            // Time
+            public void Visit(Time32Type datatype)
+            {
+                FormatString = String.Format("tt{0}", TimeUnitComponent(datatype.Unit));
+            }
+            public void Visit(Time64Type datatype)
+            {
+                FormatString = String.Format("tt{0}", TimeUnitComponent(datatype.Unit));
+            }
+            // Timestamp type
+            public void Visit(TimestampType datatype)
+            {
+                FormatString = String.Format("ts{0}:{1}", TimeUnitComponent(datatype.Unit), datatype.Timezone);
+            }
+            // Nested
+            public void Visit(ListType _) => FormatString = "+l";
+            public void Visit(StructType _) => FormatString = "+s";
+            // Dictionary
+            public void Visit(DictionaryType datatype)
+            {
+                // format string is that of the indices
+                datatype.IndexType.Accept(this);
+            }
+
+            // Unsupported
+            public void Visit(IArrowType type)
+            {
+                throw new NotImplementedException($"Exporting {type.Name} not implemented");
+            }
+        }
+
+        private sealed class ImportedArrowSchema : IDisposable
+        {
+            private readonly CArrowSchema _data;
+            private readonly IntPtr _handle;
+            private readonly bool _isRoot;
+
+            public ImportedArrowSchema(IntPtr handle)
+            {
+                _data = Marshal.PtrToStructure<CArrowSchema>(handle);
+                if (_data.release == null)
+                {
+                    throw new Exception("Tried to import a schema that has already been released.");
+                }
+                _handle = handle;
+                _isRoot = true;
+            }
+
+            private ImportedArrowSchema(IntPtr handle, bool isRoot) : this(handle)
+            {
+                _isRoot = isRoot;
+            }
+
+            public void Dispose()
+            {
+                // We only call release on a root-level schema, not child ones.
+                if (_isRoot)
+                {
+                    _data.release(_handle);
+                }
+            }
```

Comment:
`Dispose` should be idempotent.  Also, there is [a pattern](https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose) that is typically used with `Dispose()` and `Dispose(bool)` though that is probably not strictly needed here since there is no managed state.